src/Feldspar/Compiler/Compiler.hs

```diff
@@ -154,3 +167,148 @@ sicsOptions2 = defaultOptions { frontendOpts = defaultFeldOpts { targets = [SICS
 
 sicsOptions3 :: Options
 sicsOptions3 = defaultOptions { platform = c99Wool, frontendOpts = defaultFeldOpts { targets = [SICS,CSE,Wool] }}
+
+data BackendPass = BPFromCore
+                 | BPRename
+                 | BPAdapt
+                 | BPSplit
+                 | BPCompile
+                 | BPUnsplit
+  deriving (Eq, Enum, Bounded, Read, Show)
+
+data ProgOpts =
+    ProgOpts
+    { backOpts     :: Options
+    , passFileName :: String
+    , outFileName  :: String
+    , functionName :: String
+    , frontendCtrl :: PassCtrl FrontendPass
+    , backendCtrl  :: PassCtrl BackendPass
+    , printHelp    :: Bool
+    }
+
+defaultProgOpts :: ProgOpts
+defaultProgOpts =
+    ProgOpts
+    { backOpts     = defaultOptions
+    , passFileName = ""
+    , outFileName  = ""
+    , functionName = ""
+    , frontendCtrl = defaultPassCtrl
+    , backendCtrl  = defaultPassCtrl
+    , printHelp    = False
+    }
+
+program :: SyntacticFeld a => a -> IO ()
+program p = programOpts p defaultOptions
+
+programOpts :: SyntacticFeld a => a -> Options -> IO ()
+programOpts p opts = do args <- getArgs
+                        programOptsArgs p defaultProgOpts{backOpts = opts} args
+
+programOptsArgs :: SyntacticFeld a => a -> ProgOpts -> [String] -> IO ()
+programOptsArgs p opts args = do name <- getProgName
+                                 let opts0 = opts{passFileName = name ++ ".passes",
+                                                  outFileName = name,
+                                                  functionName = name}
+                                 let (opts1,_) = decodeOpts opts0 args
```

Comment:
Make decodeOpts return a single thing instead. And I think the bound variables can reside in the same let.

Reply:
Now the non-options are also used.